credentials/tls/certprovider/store.go

```diff
@@ -96,20 +96,21 @@ func (wp *wrappedProvider) Close() {
 // Close overrides the Close method of the embedded provider to avoid release the
 // already released reference.
 func (w *singleCloseWrappedProvider) Close() {
-	w.mu.Lock()
-	defer w.mu.Unlock()
-
-	w.provider.Close()
-	w.provider = closedProvider{}
+	newProvider := Provider(closedProvider{})
+	oldProvider := w.provider.Swap(&newProvider)
+	(*oldProvider).Close()
 }
 
 // KeyMaterial returns the key material sourced by the Provider.
 // Callers are expected to use the returned value as read-only.
 func (w *singleCloseWrappedProvider) KeyMaterial(ctx context.Context) (*KeyMaterial, error) {
-	w.mu.RLock()
-	defer w.mu.RUnlock()
+	return (*w.provider.Load()).KeyMaterial(ctx)
+}
 
-	return w.provider.KeyMaterial(ctx)
+// withProvider set provider to wrapper.
+func (w *singleCloseWrappedProvider) withProvider(provider Provider) *singleCloseWrappedProvider {
```

Comment:
I don't think we should use the java-style builder pattern here.  It looks like you just want a `newSingleCloseWrappedProvider(Provider)` function, anyway, based on usage?

Reply:
Apparently I was reviewing a stale diff.  Please disregard.